RES-116 Changes to temporal pooler monitor mixin and Region support

SaganBolliger opened the pull request:
Addresses https://jira.numenta.com/browse/RES-116

@scottpurdy @chetan51 It would be great to get some feedback on this.

chetan51 commented on htmresearch/support/union_temporal_pooler_monitor_mixin.py:
```diff
@@ -28,30 +28,180 @@
+      self._mmDistanceMetric = kwargs["distanceMetric"]
+      del kwargs["distanceMetric"]
+    else:
+      self._mmDistanceMetric = "pctOverlapLarger"
```
You can use `kwargs.get("distanceMetric", "pctOverlapLarger")`.

SaganBolliger replied:
Thanks. Actually looks like `kwargs.pop("distanceMetric", "pctOverlapLarger")` would be better since that also has the effect of removing the key from the dictionary.

chetan51 commented on htmresearch/support/union_temporal_pooler_monitor_mixin.py:
```diff
@@ -28,30 +28,180 @@
     super(UnionTemporalPoolerMonitorMixin, self).__init__(*args, **kwargs)
 
     self._mmResetActive = True  # First iteration is always a reset
+    self.mmClearHistory()
```
Why is this necessary?

SaganBolliger replied:
mmClearHistory() initializes the arrays in _mmData, so instead of initializing them separately in __init__(), calling mmClearHistory() from __init__() minimizes repetition.

chetan51 commented on htmresearch/support/union_temporal_pooler_monitor_mixin.py:
```diff
@@ -28,30 +28,180 @@
+    """
+    @return (list) duty cycles for union SDR bits
+    """
+    return self._mmData["UnionSDRDutyCycle"].tolist()
```
Naming convention: `UnionSDRDutyCycle` should be  `unionSDRDutyCycle`.

chetan51 commented on htmresearch/support/union_temporal_pooler_monitor_mixin.py:
```diff
@@ -279,22 +544,28 @@ def compute(self, *args, **kwargs):
+    self.getConnectedCounts(connectedCounts)
+    numConnections = numpy.sum(connectedCounts)
+
+    # import pdb; pdb.set_trace()
```
Remove commented-out code

SaganBolliger commented:
Thanks @chetan51 for your feedback. I've implemented those changes. I've also switched to using raw overlap scores rather than a normalized distance metric as per @subutai's suggestion.

@BoltzmannBrain: Is there anything that you'd like to see here that isn't already implemented?

chetan51 commented on htmresearch/support/union_temporal_pooler_monitor_mixin.py:
```diff
@@ -45,13 +52,150 @@ def __init__(self, *args, **kwargs):
     super(UnionTemporalPoolerMonitorMixin, self).__init__(*args, **kwargs)
 
     self._mmResetActive = True  # First iteration is always a reset
+    self.mmClearHistory()
```
Do you think we should put this in the base `MonitorMixin` class

SaganBolliger replied:
I don't see how it would break anything at least.

BoltzmannBrain commented on htmresearch/regions/TemporalPoolerRegion.py:
```diff
@@ -22,16 +22,26 @@
 from nupic.bindings.math import GetNTAReal
 from htmresearch.algorithms.union_temporal_pooler import UnionTemporalPooler
+from htmresearch.support.union_temporal_pooler_monitor_mixin \
+  import UnionTemporalPoolerMonitorMixin
```
Please use implicit continuation w/ parenthesis, not backslashes.

BoltzmannBrain commented:
@SaganBolliger LGTM
